src/decoder/plugins/UsfDecoderPlugin.cxx

```diff
@@ -0,0 +1,318 @@
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+#include "UsfDecoderPlugin.hxx"
+#include "../DecoderAPI.hxx"
+#include "../DecoderInternal.hxx"
+#include "tag/TagHandler.hxx"
+#include "fs/Path.hxx"
+#include "util/Domain.hxx"
+#include "Log.hxx"
+#include "util/ScopeExit.hxx"
+
+#include <usf.h>
+#include <psflib.h>
+
+static constexpr Domain usf_domain("usf");
+static constexpr unsigned USF_BUFFER_FRAMES = 2048;
+static constexpr unsigned USF_CHANNELS = 2;
+static constexpr unsigned USF_BUFFER_SAMPLES = USF_BUFFER_FRAMES*USF_CHANNELS;
+
+static void *
+stdio_fopen( const char * path )
+{
+    return fopen( path, "rb" );
+}
+
+static size_t
+stdio_fread( void *p, size_t size, size_t count, void *f )
+{
+    return fread( p, size, count, (FILE*) f );
+}
+
+static int
+stdio_fseek( void * f, int64_t offset, int whence )
+{
+    return fseek( (FILE*) f, offset, whence );
+}
+
+static int
+stdio_fclose( void * f )
+{
+    return fclose( (FILE*) f );
+}
+
+static long
+stdio_ftell( void * f )
+{
+    return ftell( (FILE*) f );
+}
+
+static constexpr psf_file_callbacks stdio_callbacks =
+{
+    "\\/:",
+    stdio_fopen,
+    stdio_fread,
+    stdio_fseek,
+    stdio_fclose,
+    stdio_ftell
+};
+
+struct UsfLoaderState
+{
+    bool enableCompare; // The _enablecompare tag is present in the file; passed to usf_set_compare
+    bool enableFIFOFull; // The _enableFIFOfull tag is present in the file; passed to usf_set_fifo_full
+    void * emu; // The emulator state
+};
+
+static int
+usf_loader(void * context, const uint8_t * exe, size_t exe_size,
+               const uint8_t * reserved, size_t reserved_size)
+{
+    struct UsfLoaderState * state = ( struct UsfLoaderState * ) context;
+
+    if ( exe && exe_size > 0 ) return -1;
+
+    return usf_upload_section( state->emu, reserved, reserved_size );
+}
+
+static int
+usf_info(void * context, const char * name, const char * value)
+{
+    struct UsfLoaderState * state = ( struct UsfLoaderState * ) context;
+    if (!value) return 0;
+
+    if ( strcmp( name, "_enablecompare" ) == 0)
+        state->enableCompare = 1;
+    else if ( strcmp( name, "_enableFIFOfull" ) == 0)
+        state->enableFIFOFull = 1;
+
+    return 0;
+}
+
+struct UsfTags {
+    double length = -1; // Track duration. -1 represents looping infinitely
+    double fade = 0;   // Fade out duration
+    bool setTags = false;  // If set, set the tags with the below variables
+    const TagHandler * tagHandler;
+    void * handler_ctx;
+};
+
+static double
+get_length_from_string(const char * string)
+{
+    size_t len = strlen(string);
+    int32_t total = 0; // Total time in milliseconds
+    int32_t finalMult = 1000; // Multiplier for the final unit. If no delimiter is used, use seconds.
+    int32_t localMult = 1; // Multiplier in the currently parsed unit (millisecond, second, minute).
+    int32_t tmp = 0; // 
```

Comment:
What does `tmp` mean? I mean, this is not a temporary value/variable - it is the one variable that is really important in this function, its lifetime spans the whole function body.